ngcc: source map flattening

petebacondarwin opened the pull request:
The library used by ngcc to update the source files (MagicString) is able
to generate a source-map but it is not able to account for any previous
source-map that the input text is already associated with.

There have been various attempts to fix this but none have been very
successful, since it is not a trivial problem to solve.

This PR contains a novel approach that is able to load up a tree of
source-files connected by source-maps and flatten them down into a single
source-map that maps directly from the final generated file to the original
sources referenced by the intermediate source-maps.

(Note that there are probably (definitely) some terrible performance characteristics in this PR. But I wanted to go for correctness before stressing about speed.)

FW-1558

filipesilva reviewed (commented):
I know you mentioned you wanted to go for correctness before speed, but can you check what the perf difference is? We shouldn't merge this yet if it makes everyone's first build 3x as long.

petebacondarwin commented:
@filipesilva absolutely!

JoostK commented:
Additionally, could this work be avoided if source map generation is disabled?

JoostK commented:
@petebacondarwin Sorry, I was a little vague. What I meant to say is that when an app is built without source maps enabled (the TS `sourcemap` option is `false` or `sourceMap` is disabled in the CLI configuration), the user is likely not interested in source maps coming from ngcc as well.  Therefore, the additional work of merging source maps in ngcc may be for nothing. I see your point when there's already an existing sourcemap, in which case not merging source maps would result in an incorrect result (probably just as broken as what is generated by ngcc as of today)

petebacondarwin commented:
Also we have to be careful in monorepo scenarios where the ngcc generated files might be used in different compilations that are configured differently. So I think assuming that there is no major performance issues it is safer to always process the source file...

petebacondarwin commented:
I have now tweaked this algorithm a bit more and I believe it is working as desired.

Regarding performance, I ran ngcc against the entire AIO `node_modules` folder (all package and all formats) and compared the time taken on master and this PR. The average time on master was 140 seconds. The average time for this PR was 149 seconds. So while there is an increase in processing time it is around only 6%, which I argue is acceptable. It is possible that we can do some further performance tuning to improve things further later on.

filipesilva commented:
Yeah 6% is pretty much within just normal variance even.

filipesilva reviewed (approved):
LGTM as far as the CLI and perf are concerned, also LGTM from a dev-infra review perspective.

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+      // For mapping [4,2] the incoming start and end are 2 and 5 (i.e. the range c, d, e, f)
+      //
+      const incomingStart = aToBmapping.originalSegment;
+      const incomingEnd = originalSegments[originalSegments.indexOf(incomingStart) + 1];
```
nit: since this can be an out of bound read, `incomingEnd` can be `undefined` here (which you have dealt with in all places, so that's all good). AFAIK TypeScript will infer the type of `incomingEnd` to be `SegmentMarker`, so including `undefined` explicitly in the type makes this a bit more safe:

```suggestion
      const incomingEnd: SegmentMarker|undefined = originalSegments[originalSegments.indexOf(incomingStart) + 1];
```

petebacondarwin replied:
I changed the code to

```
      const incomingEnd = incomingEndIndex < originalSegments.length ?
          originalSegments[incomingEndIndex] :
          undefined;
```

to avoid the out-of-bounds lookup - and this also had the benefit of adding the `| undefined` to the type automatically.

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+        continue;
+      }
+
+      // The `incomingStart` and `incomingEnd` are the `SegmentMarker`s in `B` that represent the
```
Thanks for these comments, this is super helpful!

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+      for (let bToCmappingIndex = outgoingStartIndex; bToCmappingIndex <= outgoingEndIndex;
+           bToCmappingIndex++) {
+        const bToCmapping = bSource.flattenedMappings[bToCmappingIndex];
+        if (bToCmapping) {
```
This would only be `undefined` if `bToCmappingIndex` is out of bounds wrt `bSource.flattenedMappings`, but it looks like the loop iterates only known indices of `bSource.flattenedMappings` (as `outgoingStartIndex` and `outgoingEndIndex` are directly derived from that array).

petebacondarwin replied:
The result of `findLastIndex()` can be -1, so `outgoingStartIndex` can also be -1, which would mean that the first `bToCmapping` could be `undefined`.

petebacondarwin replied:
I'll add a comment

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+/**
+ * A Mapping consists of two segment markers one in the generated source and one in the original
+ * source, which indicate the start of each segment. The end of a segment is indicated by the
+ * the first segment marker of another mapping whose start is greater or equal to this one.
```
double "the":

```suggestion
 * first segment marker of another mapping whose start is greater or equal to this one.
```

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+}
+
+/**
+ * A Mapping consists of two segment markers one in the generated source and one in the original
```
```suggestion
 * A Mapping consists of two segment markers: one in the generated source and one in the original
```

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+ * Find `item` in `items` or, if it is not found, push it to the end of the array.
+ * @param item the item to look for.
+ * @param items the collection in which to look for `item`.
+ * @returns the index of the `item` in `items`.
```
Many _items_ in this docblock 😆

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+export function parseMappings(
+    rawMap: RawSourceMap | null, sources: (SourceFile | null)[]): Mapping[] {
+  const mappings: Mapping[] = [];
+  if (rawMap !== null) {
```
nit: considering an early return here

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+  const mappings: Mapping[] = [];
+  if (rawMap !== null) {
+    const rawMappings = decode(rawMap.mappings);
+    if (rawMappings !== null) {
```
Together with an early return here would save two levels of indentation

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file_loader.ts:
```diff
@@ -0,0 +1,132 @@
+  constructor(private fs: FileSystem) {}
+
+  /**
+   * Load a source file, compute its source map, and recursively load and referenced source files.
```
```suggestion
   * Load a source file, compute its source map, and recursively load any referenced source files.
```

JoostK reviewed (commented):
I haven't been through the tests yet, but this is awesome work. Hats off to you! 👏

petebacondarwin commented on packages/compiler-cli/ngcc/src/sourcemaps/segment_marker.ts:
```diff
@@ -0,0 +1,86 @@
+}
+
+
+type Writable<T> = {
```
🤷‍♂ - it was great while it lasted.

petebacondarwin commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+      /** Any source files referenced by the raw source map associated with this source file. */
+      readonly sources: (SourceFile|null)[]) {
+    this.contents = removeSourceMapComments(contents).replace(/\n\n$/, '\n');
+    this.lines = this.contents.split('\n');
```
You're right. I might do that anyway.

petebacondarwin replied:
I did do it and I like it.

petebacondarwin commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+
+export class SourceFile {
+  /**
+   * The parsed mappings that have been flattened so that and intermediate source mappings have been
```
Me neither :-)

petebacondarwin replied:
```suggestion
   * The parsed mappings that have been flattened so that any intermediate source mappings have been
```

petebacondarwin commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,302 @@
+    const names: string[] = [];
+
+    // Ensure a mapping line array for each line in the generated source.
+    const mappings: SourceMapMappings = new Array(this.lines.length);
```
The next line fills every value. Does that not make the array non-holey?

JoostK replied:
In V8, it is not possible for an array to transition from holey to packed, even if all slots are filled. There's a lot of details to learn [in this post on the V8 blog](https://v8.dev/blog/elements-kinds), where this exact scenario is mentioned.

petebacondarwin replied:
I wonder if I can get a `PACKED_ELEMENTS` array and quickly populate it by using...

```
    const mappings: SourceMapMappings = this.lines.map(() => []);
```

JoostK replied:
I have no idea what the characteristics of that would be. At least I would expect it to result in a single allocation _and_ a packed array, so it sounds like a good approach to me.

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,312 @@
+      const outgoingEndIndex = incomingEnd !== undefined ?
+          findLastIndex(
+              bSource.flattenedMappings,
+              mapping => compareSegments(mapping.generatedSegment, incomingEnd) < 0) :
```
This looks like the only call site of `compareSegments` where the second argument is of type `SegmentMarker|undefined`, but there's a guard that `incomingEnd !== undefined` right before so we can be safe to pass `incomingEnd!` here and remove the `undefined` check from `compareSegments`.

petebacondarwin replied:
And there we not even any tests for `b` being `undefined` 🙀

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,312 @@
+
+      for (let bToCmappingIndex = outgoingStartIndex; bToCmappingIndex <= outgoingEndIndex;
+           bToCmappingIndex++) {
+        if (bToCmappingIndex !== -1) {
```
Could we set `outgoingStartIndex` to 0 if it starts out as -1? Then we can remove this check inside the loop.

petebacondarwin replied:
Great idea!

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,312 @@
+}
+
+export function computeLineLengths(str: string): number[] {
+  return (str.split(/\r?\n/)).map(s => s.length);
```
Wouldn't the optional `\r` cause all subsequent computations to be off? There's typically a `+1` to account for the `\n`, but we can't know afterwards whether there was a `\r`, right?

JoostK replied:
Also, using an `indexOf` check in a loop would avoid the allocations for the intermediate strings:

```ts
const lineLengths = [];
let offset = 0;
let newlineIndex = str.indexOf('\n');
while (newlineIndex !== -1) {
  lineLengths.push(newlineIndex - offset);
  offset = newlineIndex + 1;
  newlineIndex = str.indexOf('\n', offset);
}
return lineLengths;
```

(from the top of my head, no idea if it works ;))

petebacondarwin replied:
It turns out that the +1 is only needed as a marker that there is an end of line. It doesn't matter how many actual characters there are at the end of the line. I have updated the `Renderer` tests to prove it.

As far as the `computeLineLengths()` implementation goes... I spent ages trying out lots of different approaches, including using `indexOf`. They all seems to have some corner cases the just didn't work properly. I can't remember them all but consider, `''`, `'abc'`, `'\n'`, `'\nabc'`, `'abc\n'`. In the case of the implementation you proposed it doesn't capture the last line if there is no final `\n`. I'll add some tests to show it.

(You can probably improve it but then you are competing with the speed of the built-in `spit()` and `map()` functions, which I expect would be pretty fast - and the intermediate array can be thrown away immediately. Without performance testing I am not sure what is better - but I like the simplicity of `split` + `map`.)

JoostK replied:
Ah, I see. I realise now that mappings never refer to absolute positions from the start, only line/column numbers so that they are invariant to `\r\n` vs `\n`.

Regarding the implementation, your edge cases are indeed very valid and the extra tests are really useful. Let's keep it simple with `split` + `map` as long as there's no evidence of major perf downsides 👍

JoostK commented on packages/compiler-cli/ngcc/test/sourcemaps/source_file_loader_spec.ts:
```diff
@@ -0,0 +1,207 @@
+        expect(sourceFile.rawMap).toEqual(sourceMap);
+      });
+
+      // it('should handle missing files, etc');
```
should it? :-)

petebacondarwin replied:
yes I guess it should... (it might) ...but I should create tests for that

JoostK commented on packages/compiler-cli/ngcc/test/sourcemaps/source_file_spec.ts:
```diff
@@ -0,0 +1,212 @@
+          expect(sourceFile.flattenedMappings).toEqual([]);
+        });
+
+        // it('should be the same as non-flat mappings if there is only one level of source map', ()
```
Can this be enabled/removed?

petebacondarwin replied:
I was trying to decide how I could keep this test, now that I am not exposing the `mappings` in the `SourceFile` object. But I see that I can just run `parseMappings()` and use its output.

👍

petebacondarwin committed:
feat(ngcc): implement source-map flattening

The library used by ngcc to update the source files (MagicString) is able
to generate a source-map but it is not able to account for any previous
source-map that the input text is already associated with.

There have been various attempts to fix this but none have been very
successful, since it is not a trivial problem to solve.

This commit contains a novel approach that is able to load up a tree of
source-files connected by source-maps and flatten them down into a single
source-map that maps directly from the final generated file to the original
sources referenced by the intermediate source-maps.

JoostK reviewed (approved):
The changes in this PR LGTM, but I just realized something:

When rendering the generated statements to JavaScript code to insert into the JS bundle, any source mapping information present in the statements is _not_ retained as all statement outputs are concatenated into a single string:

https://github.com/angular/angular/blob/7659f2e24b5caf6f4c28256b12d9a2187515ff8f/packages/compiler-cli/ngcc/src/rendering/renderer.ts#L186-L191

Although this PR should address the unmapped bytes problem as reported in FW-1558 (did we actually verify that?), it does not improve the granularity of the source maps of the generated code.

petebacondarwin commented:
There are a few issues.. 

1. the previous source-map basically dropped all the references to the previous original source files, which is what the JIRA issue is primarily complaining about.
2. the mappings in that source-map were only showing the lowest resolution between the merged source maps, which results in them being of minimal use.
3. the output from ngcc compilation is indeed just a single string with no additional source-map information, which means that we do not benefit from things like template source map information.

This PR fixes both 1 and 2. I have checked both.

Number 3 is a more complex issue, and I feel that given that ngcc is only providing backward compatibility to libraries, and hopefully should only be for a limited amount of time, it may not be worth investing time in.

JoostK commented:
> This PR fixes both 1 and 2. I have checked both.

Excellent! 💯

> Number 3 is a more complex issue, and I feel that given that ngcc is only providing backward compatibility to libraries, and hopefully should only be for a limited amount of time, it may not be worth investing time in.

Agreed, just wanted to make sure this does indeed solve the unmapped bytes problem.

petebacondarwin commented:
Here is a link to a document that gives a high level explanation of what is going on here: https://docs.google.com/document/d/1x1laiDVFgJtpiw49ozwr91LG7hqohytWDiptJ93nJcE

gkalpak reviewed (approved):
Nice! +1 for the ASCII art 😛

petebacondarwin commented on packages/compiler-cli/ngcc/BUILD.bazel:
```diff
@@ -35,6 +35,7 @@ ts_library(
         "@npm//magic-string",
         "@npm//semver",
         "@npm//source-map",
+        "@npm//sourcemap-codec",
```
Yes! Good spot. We really should have a way of automating this... like generating the dependencies automatically via Bazel...

petebacondarwin replied:
added (and to the top level `package.json` since it was only available in the node_modules as a transitive dependency).

petebacondarwin commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -0,0 +1,313 @@
+ * If it is not found, then push `item` to the end of the array and returns its new index.
+ *
+ * @param item the item to look for.
+ * @param items the collection in which to look for `item`.
```
Nope :-)

petebacondarwin commented on packages/compiler-cli/ngcc/src/sourcemaps/segment_marker.ts:
```diff
@@ -0,0 +1,86 @@
+
+// The `1` is to indicate a newline character between the lines
+// Note that in the actual contents there could be more than one character -
+// e.g. \r\n - that indicate a newline but that is not important here.
```
improved the comment

petebacondarwin committed:
fixup! feat(ngcc): implement source-map flattening
